src/D2L.Bmx/OktaAuthenticator.cs

```diff
@@ -130,6 +136,114 @@ private bool TryAuthenticateFromCache(
 		return true;
 	}
 
+	private async Task<IOktaAuthenticatedClient?> TryAuthenticateWithDSSOAsync(
+		string org,
+		string user,
+		IOktaClientFactory oktaClientFactory,
+		bool experimental
+	) {
+		await using IBrowser? browser = await Browser.LaunchBrowserAsync( experimental );
+		if( browser is null ) {
+			return null;
+		}
+
+		Console.WriteLine( "Attempting to automatically login using DSSO." );
+		var cancellationTokenSource = new CancellationTokenSource( TimeSpan.FromSeconds( 10 ) );
+		var sessionIdTaskProducer = new TaskCompletionSource<string?>( TaskCreationOptions.RunContinuationsAsynchronously );
+		var userEmailTaskProducer = new TaskCompletionSource<string?>( TaskCreationOptions.RunContinuationsAsynchronously );
+		string? sessionId;
+		string? userEmail;
+
+		try {
+			var page = await browser.NewPageAsync().WaitAsync( cancellationTokenSource.Token );
+			string baseAddress = $"https://{org}.okta.com/";
+			int attempt = 1;
+
+			page.Load += ( _, _ ) => _ = GetSessionCookieAsync( cancellationTokenSource.Token );
+			page.Response += ( _, responseCreatedEventArgs ) => _ = GetOktaUserEmailAsync(
+				responseCreatedEventArgs.Response
+			);
+			await page.GoToAsync( baseAddress, timeout: 10000 );
+			sessionId = await sessionIdTaskProducer.Task.WaitAsync( cancellationTokenSource.Token );
+			userEmail = await userEmailTaskProducer.Task.WaitAsync( cancellationTokenSource.Token );
+
+			async Task GetSessionCookieAsync( CancellationToken cancellationToken ) {
+				var url = new Uri( page.Url );
+				if( url.Host == $"{org}.okta.com" ) {
+					string title = await page.GetTitleAsync();
+					if( title.Contains( "sign in", StringComparison.OrdinalIgnoreCase ) ) {
+						if( attempt < 3 && url.AbsolutePath != "/" ) {
+							attempt++;
+							await page.GoToAsync( baseAddress );
+						} else {
+							sessionIdTaskProducer.SetResult( null );
+						}
+						return;
+					}
+				}
+				var cookies = await page.GetCookiesAsync( baseAddress );
+				if( Array.Find( cookies, c => c.Name == "sid" )?.Value is string sid ) {
+					sessionIdTaskProducer.SetResult( sid );
+				}
+			}
+
+			async Task GetOktaUserEmailAsync(
+				IResponse response
+			) {
+				if( response.Url.Contains( $"{baseAddress}enduser/api/v1/home" ) ) {
+					string content = await response.TextAsync();
+					var home = JsonSerializer.Deserialize( content, JsonCamelCaseContext.Default.OktaHomeResponse );
+					if( home is not null ) {
+						userEmailTaskProducer.SetResult( home.Login );
```

Comment:
We can also get this from `sessions/me` route. Was thinking as a design pov we shouldn't create the authenticated client object unless we know it's the same person. But if that doesn't matter this can go and we just create the client then verify the user email afterwards